File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# 다리길이 = bridge_length
# 최대 하중 = weight
# 트럭 배열 = truck_weights

# 다리를 줄줄이 소세지로 건너는것이 아닌 넓은 다리를 여러 차선으로 하나씩 건넌다고 생각해야 함
# 1.접근방법
# 3가지 조건 st비어있을 때, st에 트럭을 더 추가할 수 있을 때, 그것도 아닐 때
# 한 트럭이 지나는데 소요되는시간은 다리 길이

from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    
    t_q = deque(truck_weights)
    st = deque([0] * bridge_length)
    total = 0
    while t_q:
        answer += 1
        total -= st.popleft()
        if total + t_q[0] > weight:
            st.append(0)
        else:
            w = t_q.popleft()
            total += w
            st.append(w)
            
    answer += bridge_length
            
    return answer

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Integer> t_q = new LinkedList<>();
        // 다리크기가 bridge_length인 다리List b_q
        Queue<Integer> b_q = new LinkedList<>();
        int total = 0;
        
        for (int i=0; i<truck_weights.length; i++) {
            t_q.offer(truck_weights[i]);
        }
        for (int i=0; i<bridge_length; i++) {
            b_q.offer(0);
        }
        
        while(!t_q.isEmpty()) {
            answer++;
            total -= b_q.poll();
            if (total + t_q.peek() <= weight) {
                total += t_q.peek();
                b_q.offer(t_q.poll());
            } else {
                b_q.offer(0);
            }
        }
//         마지막 대기트럭이 다리위에 올랐기 때문에 다리길이만큼 더해주면 끝
        answer += bridge_length;
        return answer;
    }
}